contracts/liquidity_hub/bonding-manager/src/rewards/commands.rs

```diff
@@ -0,0 +1,263 @@
+use cosmwasm_std::{
+    ensure, from_json, BankMsg, Coin, CosmosMsg, DepsMut, Env, MessageInfo, Reply, Response,
+    SubMsg, Uint128,
+};
+use cw_utils::parse_reply_execute_data;
+
+use white_whale_std::bonding_manager::{GlobalIndex, RewardBucket, UpcomingRewardBucket};
+use white_whale_std::epoch_manager::epoch_manager::Epoch;
+use white_whale_std::pool_network::asset;
+
+use crate::helpers::{fill_upcoming_reward_bucket, get_expiring_reward_bucket};
+use crate::state::{CONFIG, GLOBAL, LAST_CLAIMED_EPOCH, REWARD_BUCKETS, UPCOMING_REWARD_BUCKET};
+use crate::{helpers, ContractError};
+
+/// Handles the new epoch created by the epoch manager. It creates a new reward bucket with the
+/// fees that have been accrued in the previous epoch, creates a new bucket for the upcoming rewards
+/// and forwards the fees from the expiring bucket to the new one.
+pub(crate) fn on_epoch_created(
+    deps: DepsMut,
+    info: MessageInfo,
+    current_epoch: Epoch,
+) -> Result<Response, ContractError> {
+    cw_utils::nonpayable(&info)?;
+
+    // A new epoch has been created, update rewards bucket and forward the expiring bucket
+    let config = CONFIG.load(deps.storage)?;
+    ensure!(
+        info.sender == config.epoch_manager_addr,
+        ContractError::Unauthorized
+    );
+
+    let global = GLOBAL.may_load(deps.storage)?;
+    // This happens only on the very first epoch where Global has not been initialised yet
+    if global.is_none() {
+        let initial_global_index = GlobalIndex {
+            epoch_id: current_epoch.id,
+            last_updated: current_epoch.id,
+            ..Default::default()
+        };
+        GLOBAL.save(deps.storage, &initial_global_index)?;
+    }
+
+    // Update the global index epoch id
+    let mut global_index = GLOBAL.load(deps.storage)?;
```

Comment:
not exactly the same thing cuz it could be none (only the first time). You gave me an idea though, I refactored it so it's better and we don't need to load from the state twice, I remember it was a quick and dirty solution that never came back to fix.